specification.md

```diff
@@ -604,263 +514,687 @@ retries: []
+Merging number types should be done by overwriting the data from events data/action results into the merging element of the state data. 
 
-The `runBefore` property defines a name of a workflow state to be executed before workflow instance is terminated.
-States referenced by `runBefore` (as well as any other states that they transition to) must obey following rules:
-* They should not have any incoming transitions (should not be part of the main workflow control-flow logic)
-* They cannot be states marked for compensation (have their `usedForCompensation` property and set to `true`)
-* If it is a single state, it must define an [end definition](#End-Definition), if it transitions to other states,
-at last one must define it.
-* They can transition only to states are also not part of the main control flow logic (and are not marked 
-for compensation).
+Merging string types should be done by overwriting the data from events data/action results into the merging element of the state data.
+
+### Workflow Functions
+
+Workflow [functions](#Function-Definition) are reusable definitions for RESTful service invocations and/or expression evaluation.
+They can be referenced by their domain-specific names inside workflow [states](#State-Definition).
+
+Reference the following sections to learn more about workflow functions:
+* [Using functions for RESTful service invocations](#Using-Functions-For-RESTful-Service-Invocations)
+* [Using functions for RPC service invocation](#Using-Functions-For-RPC-Service-Invocations)
+* [Using functions for expression evaluations](#Using-Functions-For-Expression-Evaluation)
+
+### Using Functions For RESTful Service Invocations
+
+[Functions](#Function-Definition) can be used to describe services and their operations that need to be invoked during 
+workflow execution. They can be referenced by states [action definitions](#Action-Definition)] to clearly 
+define when the service operations should be invoked during workflow execution, as well as the data parameters
+passed to them if needed.
+
+Note that with Serverless Workflow we can also define service invocations via events.
+To learn more about that, please reference the [event definitions](#Event-Definition) section, 
+as well as the [actions definitions](#Action-Definition) [eventRef](#EventRef-Definition) property.
  
-Runtime implementations should raise compile time / parsing exceptions if any of the rules mentioned above are 
-not obeyed in the workflow definition. 
+Because of an overall lack of a common way to describe different services and their operations,
+many workflow languages typically chose to define custom function definitions.
+This approach however often runs into issues such as lack of portability, limited capabilities, as well as 
+forcing non-workflow-specific information, such as service authentication, to be added inside the workflow language.
```

Comment:
```suggestion
Because of an overall lack of a common way to describe different services and their operations,
many workflow languages typically chose to define custom function definitions.
This approach, however, often runs into issues such as lack of portability, limited capabilities, as well as 
forcing non-workflow-specific information, such as service authentication, to be added inside the workflow language.
```